src/Middleware/RequestDecompression/src/RequestDecompressionMiddleware.cs

```diff
@@ -0,0 +1,75 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Http.Metadata;
+
+namespace Microsoft.AspNetCore.RequestDecompression;
+
+/// <summary>
+/// Enables HTTP request decompression.
+/// </summary>
+internal sealed class RequestDecompressionMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly IRequestDecompressionProvider _provider;
+
+    /// <summary>
+    /// Initialize the request decompression middleware.
+    /// </summary>
+    /// <param name="next">The delegate representing the remaining middleware in the request pipeline.</param>
+    /// <param name="provider">The <see cref="IRequestDecompressionProvider"/>.</param>
+    public RequestDecompressionMiddleware(
+        RequestDelegate next,
+        IRequestDecompressionProvider provider)
+    {
+        if (next is null)
+        {
+            throw new ArgumentNullException(nameof(next));
+        }
+
+        if (provider is null)
+        {
+            throw new ArgumentNullException(nameof(provider));
+        }
+
+        _next = next;
+        _provider = provider;
+    }
+
+    /// <summary>
+    /// Invoke the middleware.
+    /// </summary>
+    /// <param name="context">The <see cref="HttpContext"/>.</param>
+    /// <returns>A task that represents the execution of this middleware.</returns>
+    public Task Invoke(HttpContext context)
+    {
+        var decompressionStream = _provider.GetDecompressionStream(context);
+        if (decompressionStream is null)
+        {
+            return _next(context);
+        }
+
+        return InvokeCore(context, decompressionStream);
+    }
+
+    private async Task InvokeCore(HttpContext context, Stream decompressionStream)
+    {
+        var request = context.Request.Body;
+        try
+        {
+            var sizeLimit =
+                context.GetEndpoint()?.Metadata?.GetMetadata<IRequestSizeLimitMetadata>()?.MaxRequestBodySize
+                    ?? context.Features.Get<IHttpMaxRequestBodySizeFeature>()?.MaxRequestBodySize;
```

Comment:
I think we should set `IHttpMaxRequestBodySizeFeature.MaxRequestBodySize read limit off of `MaxRequestBodySize` like [MVC's RequestSizeLimitFilter](https://github.com/dotnet/aspnetcore/blob/3ea008c80d5cc63de7f90ddfd6823b7b006251ff/src/Mvc/Mvc.Core/src/Filters/RequestSizeLimitFilter.cs#L60-L61)? It will be redundant in the MVC case, but we'll remove it once we address #40452. We might remove the `RequestSizeLimitFilter` as well.

Also, we should do this logic whether or not `decompressionStream is null`. It's a little weird, but once we start enforcing the request size limit to some requests to a given endpoint, we really should enforce the request size limit on all requests to that endpoint whether or not the request body was compressed.

Reply:
I've updated this to set `IHttpMaxRequestBodySizeFeature.MaxRequestBodySize` using `IRequestSizeLimitMetdata.MaxRequestBodySize`.

Reply:
> If `IRequestSizeLimitMetadata` is null, I'm assuming we should leave `IHttpMaxRequestBodySizeFeature.MaxRequestBodySize` unchanged, correct?

No. If `IRequestSizeLimitMetadata` is not present, we should leave `IHttpMaxRequestBodySizeFeature.MaxRequestBodySize` unchanged like you say because that means the endpoint has no attribute changing things one way or the other.

If `IRequestSizeLimitMetadata` is present and `IRequestSizeLimitMetadata.MaxRequestBodySize` is null, that means that `[DisableRequestSizeLimit]` or other metadata trying to do the same thing has been applied to the endpoint. In that case, we should ignore the initial default value of `IHttpMaxRequestBodySizeFeature.MaxRequestBodySize`. Instead of getting the value from the feature, we should set the value on the feature to null so the server doesn't limit the size of the request body before it gets to the decompression middleware.

In fact, we should *always* set `IHttpMaxRequestBodySizeFeature.MaxRequestBodySize` to `IRequestSizeLimitMetadata.MaxRequestBodySize` if `IRequestSizeLimitMetadata` exists. You could have a similar problem with the server limiting the size of the body before it gets to the decompression middleware even if the limit was set to a finite value like 1 GB.

Reply:
Gotcha! That's how I setup the logic in the current implementation.